docs: Add strict syntax migration guide (#4146)
nf-core/website: sites/docs/src/content/docs/developing/migration-guides/strict-syntax.md

If you see messages, see the next section for common fixes.

## Common fixes

This section describes common fixes and solutions for issues in existing `nf-core/configs`.

### Simple variables

The strict syntax does not allow variables in configs, so you will see an error like this:

```console
Error <config_name>.config:1:1: Variable declarations cannot be mixed with config statements
```

If your config uses 'simple' variables, for example static values like `scratch_dir = '/ptmp'`, follow the steps below.
For variables with conditions inside, see [Dynamic variables](#dynamic-variables).

First, check whether the variable is used multiple times.
If it isn't, remove the variable entirely and use its contents directly where the variable was being used.
If the variable is used multiple times, convert it to a parameter:

```diff groovy
- def variable_name = <code>
+ params.variable_name = <code>
```

:::warning
Make sure the parameter names are unique and isolated to the config so they don't overwrite anything in the pipelines themselves.
We recommend the format `<config_name>_<variable_name>`, but you can make it more unique if needed.
:::

> To prevent [nf-schema](https://github.com/nextflow-io/nf-schema) warnings during pipeline initialisation, add the following to your config:

Reviewer: ```suggestion
To prevent [nf-schema](https://github.com/nextflow-io/nf-schema) warnings during pipeline initialisation, add custom parameters to the `ignoreParams` list in your config:
```

```console
Error <config_name>.config:1:1: Variable declarations cannot be mixed with config statements
```

If your config uses dynamic variables, for example variables with a condition inside, follow the steps below.
For static variables, see [Simple variables](#simple-variables).

Convert these variables to parameters.
For simple cases, the code can often be converted to a ternary one-liner:

```groovy
params.random_var = something ? 'not_random' : 'random'
```

For more complicated cases, embed the conditions inside a parameter containing a closure and use `.call()` to evaluate it on config resolution.
For example, the following assignment:

```groovy
def random_var = ''
if (something == true) {
    random_var = 'not_random'
} else {
    random_var = 'random'
}
```

Becomes:

```groovy
params.random_var = {
    if (something == true) {
        return 'not_random'
    } else {
        return 'random'
    }
}.call()
```

:::warning
Don't forget to add `.call()` at the end.
This ensures the code is evaluated during config resolution.
Without `.call()`, the parameter will be a closure instead of the expected value.
:::

> To prevent [nf-schema](https://github.com/nextflow-io/nf-schema) warnings during pipeline initialisation, add the following to your config:

Reviewer: ```suggestion
To prevent [nf-schema](https://github.com/nextflow-io/nf-schema) warnings during pipeline initialisation, add custom parameters to the `ignoreParams` list in your config:
```

```groovy
validation {
    ignoreParams = [
        <list_all_used_parameters>
    ]
}
```

For a real-world example, see [nf-core/configs PR #1013](https://github.com/nf-core/configs/pull/1013/changes#diff-4adf4ac5644b18a57f91cb4b39a6d52d8f6618253441c4a61f4efa9a42a25956R95-R101).

### Functions

The strict syntax no longer allows functions in configs, so you will see an error like this:

```console
Error <config name>.config:630:14: Unexpected input: '('
│ 630 | def check_max(obj, type) {
```

Refactor your code to avoid using functions.

:::warning
There is currently no clean solution for this case because of a conflict with nf-schema, which is used for pipeline input validation.
For now, reimplement the code at each use of the function within a closure.
:::

~~All functions should be converted to callable closures that are assigned to a parameter.
For example, the following function:~~

```groovy
def calculate_something(memory, time) {
    def output = null
    // function code
    return output
}
```

> Becomes:

Reviewer: ```suggestion
~Becomes:~
```

```groovy
params.calculate_something = { memory, time ->
    def output = null
    // function code
    return output
}
```

> Calling the function can then be done via `params.calculate_something(memory, time)` instead of `calculate_something(memory, time)`.

Reviewer: ```suggestion
~Calling the function can then be done via `params.calculate_something(memory, time)` instead of `calculate_something(memory, time)`.~
```

> For a real-world example, see [nf-core/configs PR #1015](https://github.com/nf-core/configs/pull/1015/changes#diff-c60bd9c6097498d07b2f2eb3937b7d4ab3cb15e9167bacf80cb49c9848806e6fR117-R119).

Reviewer: ```suggestion
~For a real-world example, see [nf-core/configs PR #1015](https://github.com/nf-core/configs/pull/1015/changes#diff-c60bd9c6097498d07b2f2eb3937b7d4ab3cb15e9167bacf80cb49c9848806e6fR117-R119).~
```
